fix: improve thread-safety in repository implementations

divakarmanoj opened the pull request:
## Summary

Improve thread-safety in AWS S3, GCP Storage, and Git repository implementations.

## Changes

- Add `sync.Once` for lazy client initialization (AWS, GCP)
- Add `clientInitErr` field to store initialization errors
- Move network I/O outside of locks for better performance
- Improve lock scoping (only lock during data updates)
- Fix potential race condition in git repository refresh

## Test plan

- [x] Existing tests pass
- [x] No breaking changes

divakarmanoj committed:
fix: improve thread-safety in repository implementations

- Add sync.Once for lazy client initialization (AWS, GCP)
- Add clientInitErr field to store initialization errors
- Move network I/O outside of locks for better performance
- Improve lock scoping (only lock during data updates)
- Fix potential race condition in git repository refresh

Cidyuk commented on source/aws_repository.go:
```diff
@@ -15,28 +15,37 @@ import (
 }
 
 // Refresh reads the YAML file from the S3 bucket, unmarshal it into the data map.
 func (a *AwsS3Repository) Refresh() error {
```
in the client code on the refresh we have a timeout 
```
select {
		case <-ticker.C:
			// The ticker has ticked, indicating it's time to refresh the data
			err := client.Repository.Refresh() // Call the Refresh method of the repository to update the configuration data
			if err != nil {
				logrus.WithError(err).Error("error refreshing repository")
			}
		case <-ctx.Done():
			// The context is canceled, indicating the refresh routine should stop
			return
		}
```
but we are not passing the context to the refresh function and doing `ctx := context.Background()` 
that means if config.LoadDefaultConfig(ctx) will stuck - it will stuck forever or until lsome global timeout will hit - let's pass the context to refresh function so we will kill this process in case if timeout

divakarmanoj replied:
We'll do this in 2 phases:

**Phase 1 (this PR):** Apply the 'unmarshal outside lock' fix only
- No interface change
- PRs #3, #4, #5 stay independent and can be merged in any order

**Phase 2 (new PR after #3, #4, #5 merge):** Add context parameter to Refresh()
- Clean atomic change across all files
- No merge conflicts
- Can also update sardine-all's mock at the same time

This keeps the PRs independent and avoids merge conflicts between them.

Cidyuk commented on source/gcp_repository.go:
```diff
@@ -53,7 +53,7 @@ func (g *GcpStorageRepository) Refresh() error {
 	}
 
-	// Unmarshal the YAML data into the data map.
+	// Only lock for data update
```
i think we may end up in a sticky situation here 
if we will get the error during the unmarshall - it may corrupt `g.data` as we unmarshall it directly + this is an IO op so would be a good idea to move it outside the mutex

```
    // parse to temp data
    newData := make(map[string]interface{})
	if err := yaml.Unmarshal(fileContent, &newData); err != nil {
		return fmt.Errorf("yaml unmarshal failed: %w", err)
	}

	// super quick swap lock
	a.Lock()
	defer a.Unlock()
	
	a.data = newData
	a.rawData = fileContent

    return nil
```

divakarmanoj replied:
Fixed - now unmarshal happens to a temp variable outside the lock, then atomically swap the data inside the lock. This prevents data corruption if unmarshal fails.

Cidyuk commented on source/aws_repository.go:
```diff
@@ -52,7 +61,7 @@ func (a *AwsS3Repository) Refresh() error {
 	}
 
-	// Unmarshal the YAML data into the data map.
+	// Only lock for data update
```
same here as below

divakarmanoj replied:
Fixed - same pattern applied here. Unmarshal to temp variable outside lock, then atomic swap inside lock.

divakarmanoj committed:
Fix: unmarshal outside lock to prevent data corruption

- All repositories now unmarshal to temp variable outside lock
- Atomic data swap inside lock prevents corruption on unmarshal failure
- GitRepository simplified: uses sync.Once for clone, no lock for pull
- Improves performance by not holding locks during CPU-bound operations
